File: sdk/ios/Sources/FrakSDK/Core/Hex.swift
import Foundation

/// Lowercase hex, both directions. Shared by the identity proof layout and the
/// FrakContext v2 codec, which both read and emit raw bytes as hex text.
enum Hex {
    private static let digits = Array("0123456789abcdef")

    static func encode(_ bytes: some Sequence<UInt8>) -> String {
        var out = ""
        for byte in bytes {
            out.append(digits[Int(byte >> 4)])
            out.append(digits[Int(byte & 0x0F)])
        }
        return out
    }

    /// Nil on an odd length or any non-hex character.
    static func decode(_ value: String) -> Data? {
        let characters = Array(value.utf8)
        guard characters.count % 2 == 0 else { return nil }
        var bytes = Data(capacity: characters.count / 2)
        for index in stride(from: 0, to: characters.count, by: 2) {
            guard let high = nibble(characters[index]), let low = nibble(characters[index + 1]) else { return nil }
            bytes.append(high << 4 | low)
        }
        return bytes
    }

    /// The value of one ASCII hex digit, or nil.
    static func nibble(_ byte: UInt8) -> UInt8? {
        switch byte {
        case UInt8(ascii: "0")...UInt8(ascii: "9"): byte - UInt8(ascii: "0")
        case UInt8(ascii: "a")...UInt8(ascii: "f"): byte - UInt8(ascii: "a") + 10
        case UInt8(ascii: "A")...UInt8(ascii: "F"): byte - UInt8(ascii: "A") + 10
        default: nil
        }
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/core/Hex.kt
package id.frak.sdk.core

/**
 * Lowercase hex, both directions. Shared by the identity proof layout ([id.frak.sdk.identity.ProofCodec])
 * and the FrakContext v2 codec ([id.frak.sdk.sharing.FrakContextCodec]). Mirrors iOS's `Hex.swift`.
 * UUID formatting built on top of this lives in [Uuid], not here or in either codec.
 */
internal object Hex {
    private const val DIGITS = "0123456789abcdef"

    fun encode(
        bytes: ByteArray,
        offset: Int = 0,
        length: Int = bytes.size - offset,
    ): String {
        val out = StringBuilder(length * 2)
        for (index in 0 until length) {
            val byte = bytes[offset + index].toInt() and 0xFF
            out.append(DIGITS[byte ushr 4]).append(DIGITS[byte and 0xF])
        }
        return out.toString()
    }

    /** Null on an odd length or any non-hex character; never throws. */
    fun decodeOrNull(value: String): ByteArray? {
        if (value.length % 2 != 0) return null
        val out = ByteArray(value.length / 2)
        for (index in out.indices) {
            val high = nibble(value[index * 2]) ?: return null
            val low = nibble(value[index * 2 + 1]) ?: return null
            out[index] = ((high shl 4) or low).toByte()
        }
        return out
    }

    /** Writes [hex] (already validated even-length hex) into [out] at [offset]. */
    fun writeInto(
        hex: String,
        out: ByteArray,
        offset: Int,
    ) {
        for (index in 0 until hex.length / 2) {
            out[offset + index] = hex.substring(index * 2, index * 2 + 2).toInt(16).toByte()
        }
    }

    fun nibble(char: Char): Int? =
        when (char) {
            in '0'..'9' -> char - '0'
            in 'a'..'f' -> char - 'a' + 10
            in 'A'..'F' -> char - 'A' + 10
            else -> null
        }
}
